src/internal/storage/fileset/postgres_store.go

```diff
@@ -69,7 +81,54 @@ func (s *postgresStore) get(ctx context.Context, q sqlx.QueryerContext, id ID) (
 }
 
 func (s *postgresStore) Get(ctx context.Context, id ID) (*Metadata, error) {
-	return s.get(ctx, s.db, id)
+	md, err := s.getFromCache(ctx, id)
+	if err == nil {
+		return md, err
+	}
+	b := backoff.NewExponentialBackOff()
+	b.InitialInterval = 1 * time.Millisecond
+	if err := backoff.RetryUntilCancel(ctx, func() error {
+		var err error
+		md, err = s.getFromCache(ctx, id)
+		return err
+	}, b, func(err error, _ time.Duration) error {
+		if !pacherr.IsNotExist(err) {
+			return err
+		}
+		return s.deduper.Do(ctx, id, func() error {
+			md, err := s.get(ctx, s.db, id)
+			if err != nil {
+				return err
+			}
+			return s.putInCache(ctx, id, md)
+		})
+	}); err != nil {
+		return nil, err
+	}
+	return md, nil
+}
+
+func (s *postgresStore) getFromCache(ctx context.Context, id ID) (*Metadata, error) {
+	var mdData []byte
+	if err := s.cache.Get(ctx, id[:], func(data []byte) error {
+		mdData = data
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	md := &Metadata{}
+	if err := proto.Unmarshal(mdData, md); err != nil {
```

Comment:
I'm pretty sure `proto.Unmarshal` will copy the underlying bytes. json.Unmarshal does.  So you should be able to stick this in the callback, and we don't have to make another copy.